### PR TITLE
Exact dynamic MPI algorithm

### LINKED ISSUE #48: do a dynamic MPI algorithm
(for exact solving)

### PR DESCRIPTION
**Pull request by lquenti**
Closes #48

### CODE REVIEW

**Comment by lquenti** [comment]
tested it for small example, and v0 and v1 gave the same results.

There are two TODOs left where we convert between array and vector for MPI serialization, but if thats okay for you I would just leave it for now and maybe write it in our report.

File: src/solvers/exact.rs [hunk 1]
```diff
@@ -1047,6 +1045,209 @@ fn _mpi_improved_solver_rec(
     }
 }
 
+/// Calculating an exact solution parallelized using MPI and dynamic work allocation.
```
Comment (johann-cm) [inline src/solvers/exact.rs:1048]: I like the thorough explanation. It becomes very clear, what the next algorithm is doing, and why it is relevant.

File: src/solvers/exact.rs [hunk 2]
```diff
@@ -1047,6 +1045,209 @@ fn _mpi_improved_solver_rec(
...
+/// Although the root knows could already know which prefix resulted in the global minimum (by
+/// keeping track of which ones it assigned), it does not know the whole minimal path. Thus, we
+/// still need the same wrapup as the [`static_mpi_solver_generic`]:
+///
+/// After the barrier was broken the coordinator broadcasts who won and its cost.
+/// Everybody now knows who won. The winner then finally broadcasts the winning tour to every node.
+/// Thus, at the end, every node knows the best cost (from the root) and the best path (from the
+/// winner). Remember that this path-sending was done lazily as an network efficiency optimization.
+#[cfg(feature = "mpi")]
+pub fn dynamic_mpi_solver(graph_matrix: &NAMatrix) -> Solution {
```
Comment (johann-cm) [inline src/solvers/exact.rs:1108]: very clean high-level routine :+1:

File: src/solvers/exact.rs [hunk 3]
```diff
@@ -1047,6 +1045,209 @@ fn _mpi_improved_solver_rec(
...
+    let winner_process = world.process_at_rank(winner.1);
+    winner_process.broadcast_into(&mut winner_path[..]);
+
+    // After we all know the cost and path, we can finally return with the exact result
+    (winner.0, winner_path)
+}
+
+/// As it is a very interconnected logic, see [`dynamic_mpi_solver`] for a thorough explaination.
+#[cfg(feature = "mpi")]
+fn dynamic_mpi_solver_root(world: &SystemCommunicator, graph_matrix: &NAMatrix) -> MPICostRank {
```
Comment (johann-cm) [inline src/solvers/exact.rs:1143]: good documentation of *the purpose* of the lines

File: src/solvers/exact.rs [hunk 4]
```diff
@@ -1047,6 +1045,209 @@ fn _mpi_improved_solver_rec(
...
+    // Since we told all processes that there is nothing to do anymore, they all wait at the
+    // barrier. So we can break it to start the wrap up.
+    world.barrier();
+
+    current_winner
+}
+
+/// As it is a very interconnected logic, see [`dynamic_mpi_solver`] for a thorough explaination.
+#[cfg(feature = "mpi")]
+fn dynamic_mpi_solver_nonroot(world: &SystemCommunicator, graph_matrix: &NAMatrix) -> Path {
```
Comment (johann-cm) [inline src/solvers/exact.rs:1206]: good documentation of *the purpose* of the written lines

**Review by johann-cm — APPROVED** [review]
nice work, i like the documentation

**Comment by johann-cm** [comment]
> There are two TODOs left where we convert between array and vector for MPI serialization, but if thats okay for you I would just leave it for now and maybe write it in our report.

I presume that those conversions only happen a constant amount of times? Then I wold be fine with letting them stay for the time being.

**Pull request merged**